README
Pcb is a handy tool for the X Window System to design
printed circuit boards. All coordinate units are 1/1000 inch.
For details see the manual.

The only requirements are:

1. An installation of X11 including header files and the athena
   widget library (HP users that means you), GNU M4, flex
   and bison. X11R6, GNU flex and GNU bison are highly recommended.

2. An ANSI compiler (eg GCC) and maybe a newer version of bison and flex
   as a replacement for yacc and lex. See 'Problems' in the manual.

3. A 'long' filesystem


For installing the release refer to the file 'INSTALL'.
For additional information read the manual (doc.ps/pcb.ps.gz) 

If you are updating read the files in capital letters

Files:
	README.xxx             this and some other files
	doc/*                  source for manuals, texinfo...
	doc.ps/*               preformatted PostScript manuals
	example/*              some examples
	lib/*                  circuit definitions
	newlib/*               newer library format  circuit definitions
	src/*                  sources, icons, ...
	tools/*                some additional scripts...

There is also a mailing list for publishing bugs, libraries ...
See 'MAILING' in this directory.

Please mail all bug reports, suggestions, additional features and element
file that you've created to

    haceaton@aplcomm.jhuapl.edu

so they can be inserted into future releases.

-------------------------------------------------------------------------
                            COPYRIGHT

  PCB, interactive printed circuit board design
  Copyright (C) 1994, 1995, 1996, 1997  Thomas Nau
  Copyright (C) 1998, 1999, 2000, 2001, 2002 harry eaton
  Copyright (C) 2001 C. Scott Ananian
  Copyright (C) 2004 MultiGiG, Inc.

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

  Contact addresses for paper mail and Email:
  harry eaton
  6697 Buttonhole Court
  Columbia, MD 21044
  haceaton@aplcomm.jhuapl.edu


NEWS
========================================================================
Release Notes for PCB 3.0.0 (Tcl/Tk based version)
========================================================================

(To be completed)

Testing git repo implementation. . . 

========================================================================
Release Notes for PCB snapshot 20040903
========================================================================

- enable -Wall when compiling with gcc.  Many compiler warnings have
  been fixed.

- add a short man page which refers the user to the real documentation

- Added ExecuteFile(file) action which allows the execution of several
  actions contained in a text file.

- Added -script <file> flag which specifies a file containing actions
  to be executed upon startup.

- Various documentation improvements

- Protect some gcc extensions with tests for gcc.  This allows
  compilation with the IRIX c compiler.

- Add support for a background image in the drawing area which can
  help in importing a layout from a scanned image.

- Updated the ~geda library 150 and 300 mil width SOIC packages
  to remove silk on pads and to more closely follow the IPC
  recommendations

- Added support for menu accelerators.

- segfault when breaking apart an element with no element selected
  bug fixed.

- autorouter improvements

- added more 400mil DIP packages to ~geda library

- fix a trace optimizer bug which produced a segfault

- other various bug fixes


========================================================================
Release Notes for PCB snapshot 20040530
========================================================================
This has been an exciting 3 months for PCB.  Many improvements have
been made in the program as well as additions and improvements to the
footprint library.  Thanks to everyone who has provided feedback,
patches, etc.

----------------------
Additions/Improvements
----------------------
  - Expanded DRC checking
  - "Realtime" DRC checking
  - DRC settings are now stored in the board fle
  - Centroid (X-Y) and bill of materials output
  - Dynamic menus.  Users can now customize the menus via a
    site-wide, per-user, or per-project menu resource file.
  - Significant improvements to autorouting of congested designs.
  - Autorouter can now use power planes.
  - Improved appearance in autorouter output.
  - Silk on pads/pins is clipped in RS-274-X output
  - Ordering is preserved when writing output files so that diff
    may be effectively used on pcb files.
  - The autoplace feature now actually works
  - Added ToggleVisibility action
  - Most of the database is now kept in rtrees which avoids linear
    searches.  This speeds up operation significantly.
  - Added SetFlag, ClrFlag, ChangeFlag actions

---------
Bug Fixes
---------
  - pin/pad bounding box calculation fixed.  Components can
    now be placed closer to the edge of the board.
  - miter trace optimization bug.
  - Fix bug with pins which are enclosed by multiple polygons.
  - Shorts to unnamed elements and pins are handled correctly.
  - Various memory leaks fixed.
  - Fixed some bugs with a really high zoom.
  - Fixed default media selection in postscript print dialog
  - Fix some DRC bugs with square pins.

------------------
Footprint Library:
------------------
  - Added SOJ footprint family (~geda library)
  - Added CTS series 742/3/4/5/6 resistor pack footprints (~cts library)
  - Added PKG_CONNECTOR_DIL macro to let the HEADER*_1 footprints work
    (~geda library)
  - Correct the SDIP footprints (~geda library)
  - Correct some Minicircuits footprints (~minicircuits library)
  - Correct some panasonic footprints (~panasonic library)
  - Change refdes silk size for sockets (~johnstech library)
  - Increase soldermask relief on some mounting holes
    (~johnstech library)
  - Convert the COMMON_SMT_DIL_MIL and COMMON_SMT_DIL_MM macros
    used by several other libraries to the hi-res format.
  - Correct pin count on 100 pin QFP packages (100 not 72)
    (~geda library)
  - Correct pin count on HEADER60_2 (60 not 50) (~geda library)

========================================================================

New Features for 2.0

(1) New library system. A directory tree is parsed and added to the library 
selector window. This directory is specified in the config.h file as 
PCBTREEDIR and should point to a directory where any pcb user that needs to 
create library elements has write privileges. The "-libtree X" switch will 
cause pcb to use directory X instead of the installed default for the new 
library directory. You should continue to be wary of the old library 
elements, many of them are not quite right. (Just check them carefully before 
fabing to be sure there are no surprises). My intent is to create a 
repository for new elements THAT PEOPLE HAVE FABRICATED SUCCESSFULLY in order 
to grow a reliable library.

(2) Element files can now contain layers (line, arcs, polygons), vias and 
elements. The new library system reads from element files (no need to learn 
M4). The expanded capability of element files allows you to put useful 
sub-circuits into the library. For example, a QFP package with a break-out 
via pattern can be stored as a library element. Be careful with this as the 
layers are saved simply by their layer number and lose any association with a 
group (e.g. whether it is the component, or solder-side layer).

(3) Improved fascilities for creating elements from drawn objects. Now you 
can put an element in the buffer, and break it into it's consitutent parts. 
Paste this to the layout, make changes, then put it back in the buffer to 
convert to an element. You can also save a buffer's contents to an element 
file (i.e. add it to the library). Of course element files are still just 
text files so you can edit them at will. Draw the element outline on the silk 
layer (lines and arcs, no polygons), use Vias to make pins or holes, and lines 

src/set.c
/* $Id: set.c,v 1.1.1.1 2006/04/10 22:05:23 tim Exp $ */

/*
 *                            COPYRIGHT
 *
 *  PCB, interactive printed circuit board design
 *  Copyright (C) 1994,1995,1996 Thomas Nau
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *  Contact addresses for paper mail and Email:
 *  Thomas Nau, Schlehenweg 15, 88471 Baustetten, Germany
 *  Thomas.Nau@rz.uni-ulm.de
 *
 */


/* routines to update widgets and global settings
 * (except output window and dialogs)
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#ifdef HAVE_STRING_H
#include <string.h>
#endif

#include "global.h"

#include "action.h"
#include "buffer.h"
#include "compat.h"
#include "crosshair.h"
#include "control.h"
#include "data.h"
#include "draw.h"
#include "error.h"
#include "find.h"
#include "gui.h"
#include "menu.h"
#include "misc.h"
#include "output.h"
#include "set.h"
#include "undo.h"

#ifdef HAVE_LIBDMALLOC
#include <dmalloc.h>
#endif

#ifdef TCL_PCB
extern Tcl_Interp *pcbinterp;
#endif

RCSID("$Id: set.c,v 1.1.1.1 2006/04/10 22:05:23 tim Exp $");




static int mode_position = 0;
static int mode_stack[MAX_MODESTACK_DEPTH];

/* ---------------------------------------------------------------------------
 * output of cursor position
 */
void
SetCursorStatusLine (void)
{
  char text[64];

  if (Marked.status)
    sprintf (text,
	     "%-i.%02d, %-i.%02d <%-li.%02d, %-li.%02d> (%-.3fmm, %-.3fmm)",
	     Crosshair.X / 100, abs(Crosshair.X % 100), Crosshair.Y / 100,
	     abs(Crosshair.Y % 100), (Crosshair.X - Marked.X) / 100,
	     abs(Crosshair.X - Marked.X) % 100, (Crosshair.Y - Marked.Y) / 100,
	     abs(Crosshair.Y - Marked.Y) % 100,
	     COOR_TO_MM * (Crosshair.X - Marked.X),
	     COOR_TO_MM * (Crosshair.Y - Marked.Y));
  else
    sprintf (text, "%-i.%02d,%-i.%02d", Crosshair.X / 100, abs(Crosshair.X % 100),
	     Crosshair.Y / 100, abs(Crosshair.Y % 100));
  SetOutputLabel (Output.CursorPosition, text);
}

/* ---------------------------------------------------------------------------
 * output of status line
 */
void
SetStatusLine (void)
{
  char text[140];
  int length;

  if (PCB->Grid == (int) PCB->Grid)
    sprintf (text,
	     "%c %s, grid=%i.%02i:%i,%s%sline=%i.%02i, via=%i.%02i(%i.%02i),"
	     " clearance=%i.%02i, text=%i%%, buffer=#%li, name: ",
	     PCB->Changed ? '*' : ' ',
	     Settings.ShowSolderSide ? "solder" : "component",
	     (int) PCB->Grid / 100, (int) (PCB->Grid) % 100,
	     (int) Settings.GridFactor,
	     TEST_FLAG (ALLDIRECTIONFLAG,
			PCB) ? "all" : (PCB->Clipping ==
					0 ? "45" : (PCB->Clipping ==
						    1 ? "45_/" : "45\\\\_")),
	     TEST_FLAG (RUBBERBANDFLAG, PCB) ? ",R, " : ", ",
	     Settings.LineThickness / 100, Settings.LineThickness % 100,
	     Settings.ViaThickness / 100, Settings.ViaThickness % 100,
	     Settings.ViaDrillingHole / 100, Settings.ViaDrillingHole % 100,
	     Settings.Keepaway / 100, Settings.Keepaway % 100,
	     Settings.TextScale, Settings.BufferNumber + 1);
  else
    sprintf (text,
	     "%c %s, grid=%5.3fmm:%i,%s%sline=%i.%02i, via=%i.%02i(%i.%02i),"
	     " clearance=%i.%02i, text=%i%%, buffer=#%li, name: ",
	     PCB->Changed ? '*' : ' ',
	     Settings.ShowSolderSide ? "solder" : "component",
	     PCB->Grid * COOR_TO_MM, (int) Settings.GridFactor,
	     TEST_FLAG (ALLDIRECTIONFLAG,
			PCB) ? "all" : (PCB->Clipping ==
					0 ? "45" : (PCB->Clipping ==
						    1 ? "45_/" : "45\\\\_")),
	     TEST_FLAG (RUBBERBANDFLAG, PCB) ? ",R, " : ", ",
	     Settings.LineThickness / 100, Settings.LineThickness % 100,
	     Settings.ViaThickness / 100, Settings.ViaThickness % 100,
	     Settings.ViaDrillingHole / 100, Settings.ViaDrillingHole % 100,
	     Settings.Keepaway / 100, Settings.Keepaway % 100,
	     Settings.TextScale, Settings.BufferNumber + 1);

  /* append the name of the layout */
  length = sizeof (text) - 1 - strlen (text);
  strncat (text, UNKNOWN (PCB->Name), length);
  text[sizeof (text) - 1] = '\0';
  SetOutputLabel (Output.StatusLine, text);
}

/* ---------------------------------------------------------------------------
 * sets cursor grid with respect to grid offset values
 */
void
SetGrid (float Grid, Boolean align)
{
  if (Grid >= 1 && Grid <= MAX_GRID)
    {
      if (align)
	{
	  PCB->GridOffsetX =
	    Crosshair.X - (int) (Crosshair.X / Grid) * Grid + 0.5;
	  PCB->GridOffsetY =
	    Crosshair.Y - (int) (Crosshair.Y / Grid) * Grid + 0.5;
	}
      PCB->Grid = Grid;
      if (Settings.DrawGrid)
	UpdateAll ();
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
* sets new zoom factor, adapts size of viewport and centers the cursor
*/
void
SetZoom (float Zoom)
{
  int old_x, old_y;

  /* special argument of 1000 zooms to extents plus 10 percent */
  if (Zoom == 1000.0)
    {
      BoxTypePtr box = GetDataBoundingBox (PCB->Data);
      if (!box)
	return;
      Zoom =
	logf (0.011 * (float) (box->X2 - box->X1) / Output.Width) / LN_2_OVER_2;
      Zoom =
	MAX (Zoom, logf (0.011 * (float) (box->Y2 - box->Y1) / Output.Height) /
	     LN_2_OVER_2);
      Crosshair.X = (box->X1 + box->X2) / 2;
      Crosshair.Y = (box->Y1 + box->Y2) / 2;
      old_x = Output.Width / 2;
      old_y = Output.Height / 2;
    }
  else
    {
      old_x = TO_SCREEN_X (Crosshair.X);
      old_y = TO_SCREEN_Y (Crosshair.Y);
    }
  Zoom = MAX (MIN_ZOOM, Zoom);
  Zoom = MIN (MAX_ZOOM, Zoom);
  Zoom_Multiplier = 0.01/expf(Zoom * LN_2_OVER_2);

  /* redraw only if something changed */
  if (PCB->Zoom != Zoom)
    {
      PCB->Zoom = Zoom;
      RedrawZoom (old_x, old_y);
    }
}

void
RedrawZoom (Position old_x, Position old_y)
{
  ScaleOutput (Output.Width, Output.Height);
  if (CoalignScreen (old_x, old_y, Crosshair.X, Crosshair.Y))
    warpNoWhere ();

  UpdateAll ();
  /* always redraw status line (used for init sequence) */
  SetStatusLine ();
}

/* ---------------------------------------------------------------------------
 * sets a new line thickness
 */
void
SetLineSize (BDimension Size)
{
  if (Size >= MIN_LINESIZE && Size <= MAX_LINESIZE)
    {
      Settings.LineThickness = Size;
      if (TEST_FLAG (AUTODRCFLAG, PCB))
        FitCrosshairIntoGrid (Crosshair.X, Crosshair.Y);
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
 * sets a new via thickness
 */
void
SetViaSize (BDimension Size, Boolean Force)
{
  if (Force || (Size <= MAX_PINORVIASIZE &&
		Size >= MIN_PINORVIASIZE &&
		Size >= Settings.ViaDrillingHole + MIN_PINORVIACOPPER))
    {
      Settings.ViaThickness = Size;
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
 * sets a new via drilling hole
 */
void
SetViaDrillingHole (BDimension Size, Boolean Force)
{
  if (Force || (Size <= MAX_PINORVIASIZE &&
		Size >= MIN_PINORVIAHOLE &&
		Size <= Settings.ViaThickness - MIN_PINORVIACOPPER))
    {
      Settings.ViaDrillingHole = Size;
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
 * sets a keepaway width
 */
void
SetKeepawayWidth (BDimension Width)
{
  if (Width <= MAX_LINESIZE && Width >= MIN_LINESIZE)
    {
      Settings.Keepaway = Width;
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
 * sets a text scaling
 */
void
SetTextScale (Dimension Scale)
{
  if (Scale <= MAX_TEXTSCALE && Scale >= MIN_TEXTSCALE)
    {
      Settings.TextScale = Scale;
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
 * sets or resets changed flag and redraws status
 */
void
SetChangedFlag (Boolean New)
{
  if (PCB->Changed != New)
    {
      PCB->Changed = New;
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
 * sets the crosshair range to the current buffer extents 
 */
void
SetCrosshairRangeToBuffer (void)
{
  if (Settings.Mode == PASTEBUFFER_MODE)
    {
      SetBufferBoundingBox (PASTEBUFFER);
      SetCrosshairRange (PASTEBUFFER->X - PASTEBUFFER->BoundingBox.X1,
			 PASTEBUFFER->Y - PASTEBUFFER->BoundingBox.Y1,
			 PCB->MaxWidth -
			 (PASTEBUFFER->BoundingBox.X2 - PASTEBUFFER->X),
			 PCB->MaxHeight -
			 (PASTEBUFFER->BoundingBox.Y2 - PASTEBUFFER->Y));
    }
}

/* ---------------------------------------------------------------------------
 * sets a new buffer number
 */
void
SetBufferNumber (int Number)
{
  if (Number >= 0 && Number < MAX_BUFFER)
    {
      Settings.BufferNumber = Number;

      /* do an update on the crosshair range */
      SetCrosshairRangeToBuffer ();
      SetStatusLine ();
    }
}

/* ---------------------------------------------------------------------------
 * updates all widgets like status, cursor position ... on screen
 */
void
UpdateSettingsOnScreen (void)
{
  SetStatusLine ();
  SetCursorStatusLine ();
  UpdateControlPanel ();
}

void
SaveMode (void)
{
  mode_stack[mode_position] = Settings.Mode;
  if (mode_position < MAX_MODESTACK_DEPTH -1)
    mode_position++;
}

void
RestoreMode (void)
{
  if (mode_position == 0)
    {
      Message("hace: underflow of restore mode\n");
      return;
    }
  SetMode(mode_stack[--mode_position]);
}


/* ---------------------------------------------------------------------------
 * set a new mode and update X cursor
 */
void
SetMode (int Mode)
{
  static Boolean recursing = False;
  /* protect the cursor while changing the mode
   * perform some additional stuff depending on the new mode
   * reset 'state' of attached objects
   */
  if (recursing)
    return;
  recursing = True;
  HideCrosshair (True);
  addedLines = 0;
  Crosshair.AttachedObject.Type = NO_TYPE;
  Crosshair.AttachedObject.State = STATE_FIRST;
  Crosshair.AttachedPolygon.PointN = 0;
  if (PCB->RatDraw)
    {
      if (Mode == ARC_MODE || Mode == RECTANGLE_MODE ||
	  Mode == VIA_MODE || Mode == POLYGON_MODE ||
	  Mode == TEXT_MODE || Mode == INSERTPOINT_MODE ||
	  Mode == THERMAL_MODE)
	{
	  Message ("That mode is NOT allowed when drawing " "ratlines!\n");
	  Mode = NO_MODE;
	}
    }
  if (Settings.Mode == LINE_MODE && Mode == ARC_MODE &&
      Crosshair.AttachedLine.State != STATE_FIRST)
    {
      Crosshair.AttachedLine.State = STATE_FIRST;
      Crosshair.AttachedBox.State = STATE_SECOND;
      Crosshair.AttachedBox.Point1.X = Crosshair.AttachedBox.Point2.X =
	Crosshair.AttachedLine.Point1.X;
      Crosshair.AttachedBox.Point1.Y = Crosshair.AttachedBox.Point2.Y =
	Crosshair.AttachedLine.Point1.Y;
      AdjustAttachedObjects ();
    }
  else if (Settings.Mode == ARC_MODE && Mode == LINE_MODE &&
	   Crosshair.AttachedBox.State != STATE_FIRST)
    {
      Crosshair.AttachedBox.State = STATE_FIRST;
      Crosshair.AttachedLine.State = STATE_SECOND;
      Crosshair.AttachedLine.Point1.X = Crosshair.AttachedLine.Point2.X =
	Crosshair.AttachedBox.Point1.X;
      Crosshair.AttachedLine.Point1.Y = Crosshair.AttachedLine.Point2.Y =
	Crosshair.AttachedBox.Point1.Y;
      Settings.Mode = Mode;
      AdjustAttachedObjects ();
    }
  else
    {
      if (Settings.Mode == ARC_MODE || Settings.Mode == LINE_MODE)
        SetLocalRef(0, 0, False);
      Crosshair.AttachedBox.State = STATE_FIRST;
      Crosshair.AttachedLine.State = STATE_FIRST;
      if (Mode == LINE_MODE && TEST_FLAG (AUTODRCFLAG, PCB))
        {
	  SaveUndoSerialNumber ();
	  ResetFoundPinsViasAndPads (True);
	  RestoreUndoSerialNumber ();
	  ResetFoundLinesAndPolygons (True);
	  IncrementUndoSerialNumber();
	}
    }

  Settings.Mode = Mode;
  modeCursor (Mode);
  if (Mode == PASTEBUFFER_MODE)
    /* do an update on the crosshair range */
    SetCrosshairRangeToBuffer ();
  else
    SetCrosshairRange (0, 0, PCB->MaxWidth, PCB->MaxHeight);

  UpdateModeSelection ();
  recursing = False;

  /* force a crosshair grid update because the valid range
   * may have changed
   */
  MoveCrosshairRelative (0, 0);
  RestoreCrosshair (True);
}

#ifdef TCL_PCB
void
SetRouteStyle (char *name)
{
  char string [100];

  STYLE_LOOP (PCB);
  {
      if (name && strcmp (name, style->Name) == 0) {
         sprintf(string, "style use %s", style->Name);
         Tcl_Eval(pcbinterp, string);
         break;
      }
  }
  END_LOOP;
}

#else

void
SetRouteStyle (char *name)
{
  char *arg, num[10];

  STYLE_LOOP (PCB);
    {
      if (name && strcmp (name, style->Name) == 0)
	{
	  arg = &num[0];
	  sprintf (num, "%d", n + 1);
	  CallActionProc (Output.Output, "RouteStyle", NULL, &arg, 1);
	  break;
	}
    }
  END_LOOP;
}
#endif

void
SetLocalRef (Location X, Location Y, Boolean Showing)
{
  static MarkType old;
  static int count = 0;

  if (Showing)
    {
      HideCrosshair (True);
      if (count == 0)
        old = Marked;
      Marked.X = X;
      Marked.Y = Y;
      Marked.status = True;
      count++;
      RestoreCrosshair (False);
    }
  else if (count > 0)
    {
      HideCrosshair (False);
      count = 0;
      Marked = old;
      RestoreCrosshair (False);
    }
}
